app/models/product.rb

```diff
@@ -1,9 +1,12 @@
 
 class Product < ApplicationRecord
+  belongs_to :category, counter_cache: true
   has_many :orders , through: :line_items
   has_many :line_items, dependent: :restrict_with_error
   has_many :carts, through: :line_items
 
+  after_commit :set_root_category_product_count
```

Comment:
@maskedXchaotic add if condition to this callback so that is is called only if the product's category is a subcateory